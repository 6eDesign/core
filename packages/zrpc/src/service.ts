import express, { Express } from 'express';
import cors from 'cors';
import bodyParser from 'body-parser';
import responseTime from 'response-time';
import compression from 'compression';
import winston from 'winston';
import expressWinston from 'express-winston';
import { OpenAPIRegistry, OpenApiGeneratorV3 } from '@asteasolutions/zod-to-openapi';
import { createHTTPResolver, getRouteKey, RouteOptions } from './router';
import { BaseDistributedEventBus } from '@6edesign/messenger';
import { OpenAPIObject } from 'openapi3-ts/oas30';
import http from 'http';

export interface Logger {
  log: (level: string, message: string) => void;
  info: (message: string) => void;
  warn: (message: string) => void;
  error: (message: string | Error) => void;
  debug: (message: string) => void;
}

export interface ZRPCServiceOptions {
  port?: number;
  name: string;
  useCors?: boolean;
  useCompression?: boolean;
  logger?: Logger;
  eventBus?: BaseDistributedEventBus;
}

export class ZRPCService {
  public readonly app: Express;
  public readonly name: string;
  private readonly port: number;
  private readonly logger: Logger;
  private routes: RouteOptions<any, any>[] = [];
  private resolverMap: Record<string, { route: RouteOptions<any, any>; resolver: (input: any) => Promise<any> }> = {};
  private server?: http.Server;

  constructor(options: ZRPCServiceOptions) {
    this.name = options.name;
    this.port = options.port ?? 80;
    this.logger = options.logger ?? console;
    this.app = express();

    if (options.eventBus) {
      options.eventBus.listen(`${this.name}ServiceQueue`, this.handleMessage.bind(this));
    }

    if (options.useCors !== false) {
      this.app.use(cors());
    }
    if (options.useCompression !== false) {
      this.app.use(compression());
    }

    this.app.use(bodyParser.json());
    this.app.use(responseTime());

    this.app.use(expressWinston.logger({
      transports: [new winston.transports.Console()],
      format: winston.format.combine(winston.format.colorize(), winston.format.json()),
      meta: true,
      msg: 'HTTP {{req.method}} {{req.url}}',
      expressFormat: true,
      colorize: false,
    }));
  }

  public addRoute<TInput, TOutput>(
    route: RouteOptions<TInput, TOutput>,
    resolver: (input: TInput) => Promise<TOutput>
  ) {
    this.routes.push(route);
    this.resolverMap[getRouteKey(route)] = { route, resolver };

    const httpMethod = route.method ?? 'get';
    this.app[httpMethod](route.path, createHTTPResolver({ route, resolver }));
  }

  private async handleMessage({ key = '', input }: { key?: string; input?: unknown } = {}) {
    if (!key || !input) throw new Error('Invalid message format: key and input are required.');

    const { route, resolver } = this.resolverMap[key] ?? {};
    if (!resolver) {
      this.logger.error(`No resolver found for message key: ${key}`);
      throw new Error('NoResolver');
    }

    try {
      const parsedInput = route.input?.parse(input);
      await resolver(parsedInput);
    } catch (error) {
      this.logger.error(error as Error);
      throw error;
    }
  }

  public start(): Promise<void> {
    const openApiDoc = this.generateOpenAPI();
    this.app.get('/openapi.json', (req, res) => res.json(openApiDoc));
    this.app.get(['/health', '/readiness'], (req, res) => res.json({ status: 'OK' }));

    this.app.use((err: Error, req: express.Request, res: express.Response, next: express.NextFunction) => {
      this.logger.error(err.stack ?? err.message);
      res.status(500).send('Something broke!');
    });

    return new Promise((resolve) => {
      this.server = this.app.listen(this.port, () => {
        this.logger.info(`${this.name} service started on port ${this.port}`);
        resolve();
      });
    });
  }

  public stop(): Promise<void> {
    return new Promise((resolve, reject) => {
      if (this.server) {
        this.server.close((err) => {
          if (err) {
            this.logger.error(`Error stopping service: ${err.message}`);
            return reject(err);
          }
          this.logger.info(`${this.name} service stopped.`);
          resolve();
        });
      } else {
        resolve(); // No server to stop
      }
    });
  }

  public generateOpenAPI(): OpenAPIObject {
    const registry = new OpenAPIRegistry();

    this.routes.forEach((route) => {
      registry.registerPath({
        method: route.method ?? 'get',
        path: route.path,
        ...route.openapi, // Spread the openapi property here
        request: {
          ...(route.method === 'get' && { query: route.input as any }),
          ...(route.method !== 'get' && {
            body: {
              description: '',
              content: { 'application/json': { schema: route.input as any } },
            },
          }),
        },
        responses: {
          200: {
            description: 'Success',
            content: {
              'application/json': {
                schema: route.output as any,
              },
            },
          },
        },
      });
    });

    const generator = new OpenApiGeneratorV3(registry.definitions);
    return generator.generateDocument({
      info: { title: `${this.name} API Spec`, version: '1.0' },
      servers: [{ url: `/` }],
      openapi: '3.0.0',
    });
  }
}