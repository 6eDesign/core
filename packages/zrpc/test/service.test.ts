import { describe, it, expect, beforeAll, afterAll } from 'vitest';
import request from 'supertest';
import { service } from './fixtures/service';
import { STATIC_PRODUCTS } from './fixtures/data';

describe('ZRPCService', () => {
  beforeAll(async () => {
    await service.start();
  });

  afterAll(async () => {
    await service.stop();
  });

  it('accepts requests and filters products', async () => {
    const resp = await request(service.app).get('/').query({ query: 'product a' }).expect(200);
    expect(resp.body).toEqual([
      {
        id: '1',
        name: 'Product A',
        price: 10.99,
      },
    ]);
  });

  it('validates input for GET requests', async () => {
    const resp = await request(service.app).get('/').expect(400);
    expect(resp.body).toEqual({
      errors: [
        {
          code: 'invalid_type',
          expected: 'string',
          received: 'undefined',
          path: ['query'],
          message: 'Required',
        },
      ],
    });
  });

  it('accepts POST requests and adds a new product', async () => {
    const newProductData = { name: 'Product C', price: 30.00 };
    const resp = await request(service.app).post('/products').send(newProductData).expect(200);
    expect(resp.body).toEqual({
      id: '3',
      name: 'Product C',
      price: 30.00,
    });
  });

  it('validates input for POST requests', async () => {
    const resp = await request(service.app).post('/products').send({ name: 'Product D' }).expect(400);
    expect(resp.body).toEqual({
      errors: [
        {
          code: 'invalid_type',
          expected: 'number',
          received: 'undefined',
          path: ['price'],
          message: 'Required',
        },
      ],
    });
  });

  it('should expose the OpenAPI spec at /openapi.json', async () => {
    const resp = await request(service.app).get('/openapi.json').expect(200);
    expect(resp.body).toMatchSnapshot();
  });
});