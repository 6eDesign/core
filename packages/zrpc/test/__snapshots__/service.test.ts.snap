// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ZRPCService > should expose the OpenAPI spec at /openapi.json 1`] = `
{
  "components": {
    "parameters": {},
    "schemas": {},
  },
  "info": {
    "title": "test-service API Spec",
    "version": "1.0",
  },
  "openapi": "3.0.0",
  "paths": {
    "/": {
      "get": {
        "parameters": [
          {
            "in": "query",
            "name": "query",
            "required": true,
            "schema": {
              "type": "string",
            },
          },
        ],
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "items": {
                    "properties": {
                      "id": {
                        "type": "string",
                      },
                      "name": {
                        "type": "string",
                      },
                      "price": {
                        "type": "number",
                      },
                    },
                    "required": [
                      "id",
                      "name",
                      "price",
                    ],
                    "type": "object",
                  },
                  "type": "array",
                },
              },
            },
            "description": "Success",
          },
        },
      },
    },
    "/products": {
      "post": {
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "properties": {
                  "name": {
                    "type": "string",
                  },
                  "price": {
                    "type": "number",
                  },
                },
                "required": [
                  "name",
                  "price",
                ],
                "type": "object",
              },
            },
          },
          "description": "",
        },
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "properties": {
                    "id": {
                      "type": "string",
                    },
                    "name": {
                      "type": "string",
                    },
                    "price": {
                      "type": "number",
                    },
                  },
                  "required": [
                    "id",
                    "name",
                    "price",
                  ],
                  "type": "object",
                },
              },
            },
            "description": "Success",
          },
        },
      },
    },
  },
  "servers": [
    {
      "url": "/",
    },
  ],
}
`;
